controls/views.py

```diff
@@ -1942,86 +1942,119 @@ def controls_selected_export_xacta_xslx(request, system_id):
-    if request.user.has_perm('view_system', system):
-        # Retrieve primary system Project
-        # Temporarily assume only one project and get first project
-        project = system.projects.first()
-        parameter_values = project.get_parameter_values(catalog_key)
-        catalog = Catalog(catalog_key, parameter_values=parameter_values)
-        cg_flat = catalog.get_flattened_controls_all_as_dict()
-        # If control id does not exist in catalog
-        if cl_id.lower() not in cg_flat:
-            return render(request, "controls/detail.html", {"catalog": catalog, "control": {}})
-
-        # Get and return the control
-        # Retrieve any related Implementation Statements filtering by control, and system.root_element, Catalog, Type
-        impl_smts = Statement.objects.filter(sid=cl_id, consumer_element=system.root_element, sid_class=catalog_key, statement_type=StatementTypeEnum.CONTROL_IMPLEMENTATION.name).order_by('pid')
-        # Retrieve Legacy Implementation Statements
-        impl_smts_legacy = Statement.objects.filter(sid=cl_id, consumer_element=system.root_element, sid_class=catalog_key, statement_type=StatementTypeEnum.CONTROL_IMPLEMENTATION_LEGACY.name)
 
-        return project, catalog, cg_flat, impl_smts, impl_smts_legacy
+def get_statement_by_component(system, control_id, catalog_key):
+    """
+    Given a system element, a control ID and a catalog key, return the associated
+    statements.
+    """
+    # Get and return the control
+    # Retrieve any related Implementation Statements filtering by control, and system.root_element, Catalog, Type
+    stmts = Statement.objects.filter(
+        sid=control_id,
+        consumer_element=system.root_element,
+        sid_class=catalog_key,
+        statement_type=StatementTypeEnum.CONTROL_IMPLEMENTATION.name
+    ).order_by('pid')
+
+    st = {}
+    for s in stmts:
+        st[s.producer_element.name] = {
+            'body': s.body,
+            'inheritance': s.inheritance,
+            'sid': s.id,
+            'producer_element_name': s.producer_element.name,
+            'producer_element_id': s.producer_element.id,
+            'status': s.status,
+            'href': reverse('control_editor_statement',
+                args=[system.id, catalog_key, control_id, s.id])
+        }
+    if st:
+        statements = dict(sorted(st.items()))
+    else:
+        statements = {}
+
+    return statements
+
+
+def get_narrative(statements, statement_id):
+    narrative = None
+    # Determine the next item in the dictionary
+    next = None
+    if statements and statement_id:
+        get_next = False
+        for k, s in statements.items():
+            if get_next:
+                next = k
+                get_next = False
+            if s['sid'] == int(statement_id):
+                narrative = s
+                statements[k]['active'] = True
+                get_next = True
+    elif statements:
+        k = list(statements.keys())[0]
```

Comment:
Hmmm.  So this is really odd, although I believe it actually might work in Python >= 3.7.  Previously the ordering of keys in a dictionary was not guaranteed, so `list(foo.keys())[0]` is kind of asking for trouble.  Even so, I'm not positive I get what is going on here.  Aren't statements constructed by `get_statement_by_component`?  So don't we know what the keys are?  Maybe I'm really missing something here :-)

Reply:
Yes, `get_statements_by_component` constructs the statements and then sorts the keys alphabetically. In this function we are trying to determine which Narrative to display. If there is a statement_id then that is the statement to display, and given the dict of statements, we want to know if there is one that comes after this one so that the narrative form button will display `Save & Next` and, after submitting the form, redirect you to the next Component  in the Statements dict. If there is no statement_id, we want to grab the first item in the sorted dict and use that for the Narrative form and, if there are more than one item in the dict, display the ...Next button. I'd love to discuss and easier/better way to achieve this.